src/ServiceControl.Infrastructure/AsyncTimer.cs

```diff
@@ -1,96 +1,106 @@
-                        }
-                        catch (OperationCanceledException) when (token.IsCancellationRequested)
+                        var result = await callback(token).ConfigureAwait(false);
+
+                        consecutiveFailures = 0;
+                        if (result == TimerJobExecutionResult.DoNotContinueExecuting)
                         {
-                            break;
+                            tokenSource.Cancel();
                         }
-                        catch (Exception ex)
+                        else if (result == TimerJobExecutionResult.ScheduleNextExecution)
                         {
-                            errorCallback(ex);
+                            await Task.Delay(interval, token).ConfigureAwait(false);
                         }
+
+                        //Otherwise execute immediately
                     }
-                }
-                catch (OperationCanceledException) when (token.IsCancellationRequested)
-                {
-                    // no-op
-                }
-            }, CancellationToken.None);
-        }
+                    catch (OperationCanceledException) when (token.IsCancellationRequested)
+                    {
+                        break;
+                    }
+                    catch (Exception ex)
+                    {
+                        consecutiveFailures++;
+                        var exponentialBackoffDelay = TimeSpan.FromSeconds(int.Max(60, consecutiveFailures * consecutiveFailures));
 
-        public async Task Stop()
-        {
-            if (tokenSource == null)
+                        await Task.Delay(exponentialBackoffDelay, token).ConfigureAwait(false);
+
+                        errorCallback(ex);
+                    }
+                }
+            }
+            catch (OperationCanceledException) when (token.IsCancellationRequested)
             {
-                return;
+                // no-op
             }
+        }, CancellationToken.None);
+    }
 
-            await tokenSource.CancelAsync().ConfigureAwait(false);
-            tokenSource.Dispose();
+    public async Task Stop(CancellationToken cancellationToken)
+    {
+        if (tokenSource == null)
+        {
+            return;
+        }
 
-            if (task != null)
-            {
-                try
-                {
-                    await task.ConfigureAwait(false);
-                }
-                catch (OperationCanceledException) when (tokenSource.IsCancellationRequested)
-                {
-                    //NOOP
-                }
-            }
+        await tokenSource.CancelAsync().ConfigureAwait(false);
+        tokenSource.Dispose();
+
+        if (task == null)
+        {
+            return;
         }
 
-        Task task;
-        CancellationTokenSource tokenSource;
+        try
+        {
+            Task.WaitAll([task], cancellationToken);
```

Comment:
Is there a better way to do this?

Reply:
thinking out laud, isn't there a `task.Wait()` thing?

Reply:
Duh, I looked for it in the wrong place 😬 

Fixed now